### PR TITLE
QUIC: Uses unidirectional stream to send HTTP3 SETTINGS frames

### PR DESCRIPTION
**Pull request by scw00**

### CODE REVIEW

File: iocore/net/quic/QUICTypes.h [hunk 1]
```diff
@@ -424,13 +424,20 @@ class QUICPreferredAddress
   bool _valid = false;
 };
 
-enum class QUICStreamType : uint8_t {
+enum class QUICStreamInitType : uint8_t {
   CLIENT_BIDI = 0x00,
```
Comment (scw00) [inline iocore/net/quic/QUICTypes.h]: Not sure what name should be used here?

File: iocore/net/quic/QUICUnidirectionalStream.cc [hunk 2]
```diff
@@ -401,7 +403,7 @@ QUICSendStream::largest_offset_sent() const
 //
 QUICReceiveStream::QUICReceiveStream(QUICRTTProvider *rtt_provider, QUICConnectionInfoProvider *cinfo, QUICStreamId sid,
                                      uint64_t recv_max_stream_data)
-  : QUICStreamVConnection(cinfo, sid),
+  : QUICStreamVConnection(cinfo, sid, QUICStreamType::RECEIVE_ONLY),
```
Comment (maskit) [inline iocore/net/quic/QUICUnidirectionalStream.cc]: Why do you want to pass RECEIVE_ONLY? You can determine it by reading a stream id.

Comment (maskit) [reply iocore/net/quic/QUICUnidirectionalStream.cc]: Passing send/receive/both would make sense if we had a stream factory that internally assigns a stream id.

For example:
```cpp
QUICStream *stream = stream_factory.create(RECEIVE);
---
QUICStream *
QUICStreamFactory::create(QUICStreamDicrection direction) {
  QUICStream stream;

  QUICStreamId id = this->_get_next_stream_id(direction);
  switch (direction) {
  case RECEIVE:
    stream = new QUICReceiveStream(this->cinfo, id);
  }
  return stream;
}
```

File: iocore/net/quic/QUICTypes.h (hunk 1 above)
Comment (maskit) [reply iocore/net/quic/QUICTypes.h]: I'd keep this as is and name the other QUICStreamDirection.

File: iocore/net/quic/QUICUnidirectionalStream.cc (hunk 2 above)
Comment (scw00) [reply iocore/net/quic/QUICUnidirectionalStream.cc]: That make sense. will do it.

Comment (scw00) [reply iocore/net/quic/QUICUnidirectionalStream.cc]: update ! can you please review again.

File: iocore/net/quic/QUICStreamFactory.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+#include "QUICTypes.h"
+
+class QUICStreamVConnection;
+
+// PS: this class function should not static because of  THREAD_ALLOC and THREAD_FREE
+class QUICStreamFactory
+{
+public:
+  // stream. negative == true means this function called by incoming frame. If the stream_id indicate an unidirectional stream,
```
Comment (maskit) [inline iocore/net/quic/QUICStreamFactory.h]: We don't have `negative` now, right?

File: iocore/net/quic/QUICStreamFactory.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  // then create a send only stream, otherwise (negtive == false) create a send only stream.
+  QUICStreamVConnection *create(QUICRTTProvider *rtt_provider, QUICConnectionInfoProvider *cinfo, QUICStreamId sid,
```
Comment (maskit) [inline iocore/net/quic/QUICStreamFactory.h]: We should probably be able to pass `rtt_provider` and `cinfo` to the constructor, then we don't need to pass them on every `create()`.

**Comment by maskit** [comment]
Overall looks good to me.

I think this is just an internal implementation change but I want @masaori335 to take a look because I haven't seen H3 part for a long time.

**Comment by masaori335** [comment]
Will look tonight, but it looks like this is underlay of HTTP/3.

**Comment by maskit** [comment]
@masaori335 

> but it looks like this is underlay of HTTP/3.

I wanted you to confirm:
- Whether the stream type (unidirectional) is appropriate for SETTINGS frame
- Whether QUICStreamManager's interface (not changed) is still reasonable for H3 (and QUIC apps)

File: iocore/net/quic/Makefile.am [hunk 5]
```diff
@@ -116,7 +117,8 @@ check_PROGRAMS = \
   test_QUICTypeUtil \
   test_QUICVersionNegotiator \
   test_QUICFrameRetransmitter \
-  test_QUICAddrVerifyState
+  test_QUICAddrVerifyState \
+  test_QUICStreamFactory
```
Comment (masaori335) [inline iocore/net/quic/Makefile.am]: I might be missing something, where is source file of this check program?

File: iocore/net/quic/QUICStreamFactory.cc [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+ClassAllocator<QUICBidirectionalStream> quicBidiStreamAllocator("quicBidiStreamAllocator");
+ClassAllocator<QUICSendStream> quicSendStreamAllocator("quicSendStreamAllocator");
+ClassAllocator<QUICReceiveStream> quicReceiveStreamAllocator("quicReceiveStreamAllocator");
+
+QUICStreamVConnection *
+QUICStreamFactory::create(QUICStreamId sid, uint64_t local_max_stream_data, uint64_t remote_max_stream_data)
+{
+  QUICStreamVConnection *stream = nullptr;
+  switch (QUICTypeUtil::detect_stream_type(sid)) {
```
Comment (masaori335) [inline iocore/net/quic/QUICStreamFactory.cc]: This switch case looks same logic of `QUICTypeUtil::detect_stream_direction()`. Is there any difference?

**Comment by masaori335** [comment]
Prior this change, QUICBidirectionalStream was used for uni stream. This change it QUICUnidirectionalStream correctly. So this should not have any impact for application.

Also introducing QUICStreamDirection looks reasonable.

File: iocore/net/quic/Makefile.am (hunk 5 above)
Comment (scw00) [reply iocore/net/quic/Makefile.am]: Ops.. missing that.

File: iocore/net/quic/QUICStreamFactory.cc (hunk 6 above)
Comment (scw00) [reply iocore/net/quic/QUICStreamFactory.cc]: The `QUICStreamType` is fixed value:
See

Bits | Stream Type
-- | --
0x0 | Client-Initiated, Bidirectional
0x1 | Server-Initiated, Bidirectional
0x2 | Client-Initiated, Unidirectional
0x3 | Server-Initiated, Unidirectional

we can use it and the context to determine which direction should be used.

If we only used `QUICStreamDirection`. We would lost some information about which side we sit (client or server).
```
enum class QUICStreamDirection : uint8_t {
  UNKNOWN = 0,
  SEND,
  RECEIVE,
  BIDIRECTIONAL,
};
```

So QUICStreamDirection indicate which type of stream should be create.
QUICStreamType and context can be used to determine some of logic like transport parameters and stream manager  initialization.

File: iocore/net/quic/Makefile.am (hunk 5 above)
Comment (scw00) [reply iocore/net/quic/Makefile.am]: Fixed.

File: iocore/net/quic/QUICStreamFactory.cc (hunk 6 above)
Comment (maskit) [reply iocore/net/quic/QUICStreamFactory.cc]: I guess @masaori335 was trying to say we could do this:

```
switch(detect_stream_direction(sid, this->_info->direction())) {
case BIDIRECTIONAL:
  ...
case SEND:
  ...
case RECEIVE:
  ...
}
```

Comment (scw00) [reply iocore/net/quic/QUICStreamFactory.cc]: fixed..

Comment (masaori335) [reply iocore/net/quic/QUICStreamFactory.cc]: Thanks :+1: I thought we would make mistake easily with the complicated logic.

**Pull request merged**